src/server/worker/driver/driver.go

```diff
@@ -0,0 +1,1248 @@
+		if parts[0] == userOrUID || parts[2] == userOrUID {
+			result := &user.User{
+				Username: parts[0],
+				Uid:      parts[2],
+				Gid:      parts[3],
+				Name:     parts[4],
+				HomeDir:  parts[5],
+			}
+			if groupOrGID != "" {
+				if parts[0] == userOrUID {
+					// groupOrGid is a group
+					group, err := lookupGroup(groupOrGID)
+					if err != nil {
+						return nil, err
+					}
+					result.Gid = group.Gid
+				} else {
+					// groupOrGid is a gid
+					result.Gid = groupOrGID
+				}
+			}
+			return result, nil
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		log.Fatal(err)
+	}
+	return nil, fmt.Errorf("user %s not found", userArg)
+}
+
+func lookupGroup(group string) (_ *user.Group, retErr error) {
+	groupFile, err := os.Open("/etc/group")
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := groupFile.Close(); err != nil && retErr == nil {
+			retErr = err
+		}
+	}()
+	scanner := bufio.NewScanner(groupFile)
+	for scanner.Scan() {
+		parts := strings.Split(scanner.Text(), ":")
+		if parts[0] == group {
+			return &user.Group{
+				Gid:  parts[2],
+				Name: parts[0],
+			}, nil
+		}
+	}
+	return nil, fmt.Errorf("group %s not found", group)
+}
+
+func (d *driver) WithCtx(ctx context.Context) Driver {
+	result := &driver{}
+	*result = *d
+	result.pachClient = result.pachClient.WithCtx(ctx)
+	return result
+}
+
+func (d *driver) Jobs() col.Collection {
+	return d.jobs
+}
+
+func (d *driver) Pipelines() col.Collection {
+	return d.pipelines
+}
+
+func (d *driver) NewTaskWorker() *work.Worker {
+	return work.NewWorker(d.etcdClient, d.etcdPrefix, workNamespace(d.pipelineInfo))
+}
+
+func (d *driver) NewTaskMaster() *work.Master {
+	return work.NewMaster(d.etcdClient, d.etcdPrefix, workNamespace(d.pipelineInfo))
+}
+
+func (d *driver) GetExpectedNumWorkers() (int, error) {
+	pipelinePtr := &pps.EtcdPipelineInfo{}
+	if err := d.Pipelines().ReadOnly(d.PachClient().Ctx()).Get(d.PipelineInfo().Pipeline.Name, pipelinePtr); err != nil {
+		fmt.Printf("GetExpectedNumWorkers failed, err: %v\n", err)
+		return 0, err
+	}
+	fmt.Printf("GetExpectedNumWorkers success: %v\n", pipelinePtr.Parallelism)
+	return int(pipelinePtr.Parallelism), nil
+}
+
+func (d *driver) NumShards() int64 {
+	return d.numShards
+}
+
+func (d *driver) PipelineInfo() *pps.PipelineInfo {
+	return d.pipelineInfo
+}
+
+func (d *driver) InputDir() string {
+	return d.inputDir
+}
+
+func (d *driver) HashtreeDir() string {
+	return d.hashtreeDir
+}
+
+func (d *driver) PachClient() *client.APIClient {
+	return d.pachClient
+}
+
+func (d *driver) ChunkCaches() cache.WorkerCache {
+	return d.chunkCaches
+}
+
+func (d *driver) ChunkStatsCaches() cache.WorkerCache {
+	return d.chunkStatsCaches
+}
+
+// This is broken out into its own function because its scope is small and it
+// can easily be used by the mock driver for testing purposes.
+func withDatumCache(storageRoot string, cb func(*hashtree.MergeCache, *hashtree.MergeCache) error) (retErr error) {
```

Comment:
It seems like it would make more sense for the directory setup and cleanup to be behind the `MergeCache` abstraction.